src/query/ts/m3db/encoded_series_iterator.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/models"
+	xts "github.com/m3db/m3/src/query/ts"
+)
+
+type encodedSeriesIter struct {
+	mu           sync.RWMutex
+	idx          int
+	meta         block.Metadata
+	bounds       models.Bounds
+	seriesMeta   []block.SeriesMeta
+	seriesIters  []encoding.SeriesIterator
+	consolidator singleConsolidator
+}
+
+func (b *encodedBlock) seriesIter() block.SeriesIter {
+	cs := b.consolidation
+	bounds := cs.bounds
+	consolidator := buildSingleConsolidator(
+		time.Minute,
+		bounds.StepSize,
+		cs.currentTime,
+		cs.consolidationFn,
+	)
+	return &encodedSeriesIter{
+		idx:          -1,
+		meta:         b.meta,
+		bounds:       bounds,
+		seriesMeta:   b.seriesMetas,
+		seriesIters:  b.seriesBlockIterators,
+		consolidator: consolidator,
+	}
+}
+
+func (it *encodedSeriesIter) Current() (block.Series, error) {
+	it.mu.RLock()
+	iter := it.seriesIters[it.idx]
+	values := make([]float64, it.bounds.Steps())
+	xts.Memset(values, math.NaN())
+	i := 0
```

Comment:
you can avoid keeping this by values := make([]float64, 0, it.bounds.Steps()) and later just appending to values

Reply:
Unfortunately that would make it awkward to fill up the values with nans if the consolidator is empty later on